File: cpp/src/detector.cpp
/// VisionTrack — ONNX Runtime Detector (C++)

#include "detector.hpp"
#include <iostream>
#include <stdexcept>

namespace vt {

Detector::Detector(const std::string& model_path,
                   int input_size,
                   int num_classes,
                   float conf_threshold,
                   float iou_threshold,
                   bool use_cuda)
    : input_size_(input_size),
      num_classes_(num_classes),
      conf_threshold_(conf_threshold),
      iou_threshold_(iou_threshold),
      env_(ORT_LOGGING_LEVEL_WARNING, "VisionTrack"),
      session_(env_, model_path.c_str(), Ort::SessionOptions{})
{
    // Get input name and shape
    auto* input = session_.GetInputInfo(0);
    input_name_ = input->GetName();

    auto* type_info = input->GetTypeInfo();
    auto* tensor_info = type_info->GetTensorTypeAndShapeInfo();
    input_shape_ = tensor_info->GetShape();

    std::cout << "Model loaded: " << model_path << "\n"
              << "  Input: " << input_name_ << " [";
    for (size_t i = 0; i < input_shape_.size(); ++i) {
        std::cout << input_shape_[i] << (i + 1 < input_shape_.size() ? "," : "");
    }
    std::cout << "]\n"
              << "  Output: " << session_.GetOutputCount() << " tensor(s)\n";

    // Update input_size from model if dynamic
    if (input_shape_.size() == 4 && input_shape_[2] > 0) {
        input_size_ = static_cast<int>(input_shape_[2]);
    }
}

std::vector<float> Detector::infer(const std::vector<float>& tensor,
                                    int channels, int height, int width)
{
    std::vector<int64_t> shape = {1, channels, height, width};
    auto memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator, OrtMemTypeDefault);

    auto input_tensor = Ort::Value::CreateTensor<float>(
        memory_info, const_cast<float*>(tensor.data()), tensor.size(),
        shape.data(), shape.size());

    const char* input_name = input_name_.c_str();
    auto output_tensors = session_.Run(
        Ort::RunOptions{}, &input_name, &input_tensor, 1, nullptr, 0);

    // Extract output data
    auto& out = output_tensors[0];
    auto* out_data = out.GetTensorMutableData<float>();
    auto out_shape = out.GetTensorTypeAndShapeInfo().GetShape();
    size_t out_size = 1;
    for (auto d : out_shape) out_size *= d;

    return std::vector<float>(out_data, out_data + out_size);
}

std::vector<Detection> Detector::postprocess(const std::vector<float>& output) {
    return decode_detections(output.data(), static_cast<int>(output.size()),
                             num_classes_, conf_threshold_, iou_threshold_);
}

}  // namespace vt

File: python/detector.py
"""
VisionTrack - ONNX Detector

Wrapper around ONNX Runtime for YOLO detection inference.
Handles model loading, session configuration, and inference.
"""

from pathlib import Path
from typing import Optional

import numpy as np
import onnxruntime as ort

from utils.postprocessing import decode_detections


class Detector:
    """ONNX Runtime YOLO detector."""

    def __init__(
        self,
        model_path: str,
        input_size: int = 640,
        num_classes: int = 80,
        conf_threshold: float = 0.25,
        iou_threshold: float = 0.45,
        class_names: Optional[list[str]] = None,
        use_cuda: bool = False,
        gpu_id: int = 0,
    ):
        """
        Args:
            model_path: Path to ONNX model
            input_size: Model input size (square)
            num_classes: Number of classes in model
            conf_threshold: Default confidence threshold
            iou_threshold: NMS IoU threshold
            class_names: Optional list of class names
            use_cuda: Whether to use CUDA execution provider
            gpu_id: GPU index
        """
        self.model_path = str(model_path)
        self.input_size = input_size
        self.num_classes = num_classes
        self.conf_threshold = conf_threshold
        self.iou_threshold = iou_threshold
        self.class_names = class_names or []

        # Inspect model first to determine num_classes dynamically
        try:
            session_options = ort.SessionOptions()
            session_options.graph_optimization_level = (
                ort.GraphOptimizationLevel.ORT_ENABLE_ALL
            )

            providers = ["CPUExecutionProvider"]
            if use_cuda:
                providers = [
                    ("CUDAExecutionProvider", {"device_id": gpu_id}),
                    "CPUExecutionProvider",
                ]

            self.session = ort.InferenceSession(
                self.model_path,
                sess_options=session_options,
                providers=providers,
            )

            # Auto-detect input size from model metadata
            input_info = self.session.get_inputs()[0]
            shape = input_info.shape
            if len(shape) == 4 and isinstance(shape[2], int):
                self.input_size = shape[2]
            self.input_name = input_info.name
            self.input_shape = shape

            # Auto-detect output
            self.output_names = [o.name for o in self.session.get_outputs()]
            self.output_shape = self.session.get_outputs()[0].shape
            print(
                f"Model loaded: {Path(self.model_path).name}\n"
                f"  Input: {input_info.name} {self.input_shape}\n"
                f"  Outputs: {self.output_names}\n"
                f"  Providers: {self.session.get_providers()}"
            )
        except Exception as e:
            raise RuntimeError(f"Failed to load ONNX model: {e}")

    def infer(self, tensor: np.ndarray) -> np.ndarray:
        """Run inference on a preprocessed input tensor."""
        outputs = self.session.run(self.output_names, {self.input_name: tensor})
        return outputs[0]

    def postprocess(
        self,
        output: np.ndarray,
        conf_threshold: Optional[float] = None,
        iou_threshold: Optional[float] = None,
    ) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        """
        Decode raw model output into (boxes_xyxy, scores, class_ids).

        Args:
            output: Raw ONNX output
            conf_threshold: Override confidence threshold
            iou_threshold: Override NMS IoU threshold

        Returns:
            Boxes in xyxy format (in model input space), scores, class ids
        """
        conf = conf_threshold if conf_threshold is not None else self.conf_threshold
        iou = iou_threshold if iou_threshold is not None else self.iou_threshold

        return decode_detections(
            output,
            num_classes=self.num_classes,
            conf_threshold=conf,
            iou_threshold=iou,
            input_size=self.input_size,
        )